# Tutorial 1 Homework Task (1.5 Marks)

This task is based on Flights data  which comes a part of nycflights13 library. Read about data from the following sources:

https://pypi.org/project/nycflights13/

https://github.com/tidyverse/nycflights13

### <span style="color:red">IMPORTANT:</span> You need to study and understand your data very carefully before answering the questions.

### Data Acquisition

In [3]:
!pip install nycflights13


In [17]:
from nycflights13 import flights
from nycflights13 import weather
from nycflights13 import airports
from nycflights13 import planes
from nycflights13 import airlines
import pandas as pd
#printing first two rows from flights data
print(flights.head(2))
#getting only required fields/columns from flights in your dataframe

flight_data=flights[['year', 'month', 'day', 'dep_time', 'sched_dep_time', 'dep_delay', 'hour', 'minute']].copy()
print('flight data')
print(flight_data.head(2))
flight_data.info()


   year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   

   distance  hour  minute             time_hour  
0      1400     5      15  2013-01-01T10:00:00Z  
1      1416     5      29  2013-01-01T10:00:00Z  
flight data
   year  month  day  dep_time  sched_dep_time  dep_delay  hour  minute
0  2013      1    1     517.0             515        2.0     5      15
1  2013      1    1     533.0             529        4.0     5      29
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 336776 entries, 0 to 336775
Data columns (total 8 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --

### <span style="color:red">Tutorial Exercise</span>

As can be noticed from flight_data.info() output, there are multiple columns with missing data. Answer the following:

a) Write python code to compute and print % of missing values in each column of 'flight_data' dataframe.

b) Which method you will utilise to handle missing values for each column? You need to give justification for your answer.

c) Give python code for the methods specified in part(b) for handling missing values.



###### <span style="color:red">Submission: </span> Mylo Assignment Box (Tutorial Task)
###### <span style="color:red">Submission Deadline: </span> within 5 Days of your Tutorial

### Preprocessing Steps

#### Answer to (a)

In [15]:
#Python code for (a)

#checking how many missing values, null in each column of 'flight_data' 
print(flight_data.isnull().sum())

#print the % of missing values
missing_pct = round(flight_data.isnull().sum()/len(flight_data) * 100, 1)
print(missing_pct)

year                 0
month                0
day                  0
dep_time          8255
sched_dep_time       0
dep_delay         8255
hour                 0
minute               0
dtype: int64
year              0.0
month             0.0
day               0.0
dep_time          2.5
sched_dep_time    0.0
dep_delay         2.5
hour              0.0
minute            0.0
dtype: float64


In [25]:
flights.isnull().sum()

year                 0
month                0
day                  0
dep_time          8255
sched_dep_time       0
dep_delay         8255
arr_time          8713
sched_arr_time       0
arr_delay         9430
carrier              0
flight               0
tailnum           2512
origin               0
dest                 0
air_time          9430
distance             0
hour                 0
minute               0
time_hour            0
dtype: int64

#### Answer to (b)

There are 8255 flights with missing dep_time. For the flights with missing dep_time -- dep_delay, arr_time, arr_delay, and air_time are also missing. By applying domain knowledge, we can see that these are the flights that were cancelled.

Using my domain knowldedge, I would delete those cells because I know the flights did not even depart, hence there will not be departure time (dep_time) or departure delay (dep_delay).  

However, for the purposes of this exercise, i will use imputation to handle missing values. more specifically, i will use timeseries imputation such as iterative imputation (regression) as (1) the flight data should have some relation between other columns and the column where there are missing values. And (2) the flight data should have trends during the periods of the flight.   

#### Answer to (c)

In [26]:
df = pd.DataFrame(flight_data)
print(df.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 336776 entries, 0 to 336775
Data columns (total 8 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   year            336776 non-null  int64  
 1   month           336776 non-null  int64  
 2   day             336776 non-null  int64  
 3   dep_time        328521 non-null  float64
 4   sched_dep_time  336776 non-null  int64  
 5   dep_delay       328521 non-null  float64
 6   hour            336776 non-null  int64  
 7   minute          336776 non-null  int64  
dtypes: float64(2), int64(6)
memory usage: 20.6 MB
None


In [31]:
#Python code for (c)

from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
impute_it = IterativeImputer()
impute_it.fit_transform(flight_data)
X1=impute_it.fit_transform(flight_data)
flight_data_adv=pd.DataFrame(X1)
#print original dataframe
print(flight_data)
#print imputed dataframe
print(flight_data_adv)
#check for missing values after iterative imputation
print(flight_data_adv.isnull().sum())

        year  month  day  dep_time  sched_dep_time  dep_delay  hour  minute
0       2013      1    1     517.0             515        2.0     5      15
1       2013      1    1     533.0             529        4.0     5      29
2       2013      1    1     542.0             540        2.0     5      40
3       2013      1    1     544.0             545       -1.0     5      45
4       2013      1    1     554.0             600       -6.0     6       0
...      ...    ...  ...       ...             ...        ...   ...     ...
336771  2013      9   30       NaN            1455        NaN    14      55
336772  2013      9   30       NaN            2200        NaN    22       0
336773  2013      9   30       NaN            1210        NaN    12      10
336774  2013      9   30       NaN            1159        NaN    11      59
336775  2013      9   30       NaN             840        NaN     8      40

[336776 rows x 8 columns]
             0    1     2            3       4          5    